maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java

```diff
@@ -266,16 +302,49 @@ private DependencyResolutionResult resolveDependencies( MavenProject project, Re
         return ids;
     }
 
-    private ModelBuildingRequest getModelBuildingRequest( InternalConfig config )
+    private ModelBuildingRequest getModelBuildingRequest( InternalConfig config, File pomFile )
     {
         ProjectBuildingRequest configuration = config.request;
 
         ModelBuildingRequest request = new DefaultModelBuildingRequest();
 
         RequestTrace trace = RequestTrace.newChild( null, configuration ).newChild( request );
 
+        RepositorySystemSession repoSession;
+        if ( Features.buildConsumer().isActive() )
+        {
+            TransformerContext context = new TransformerContext()
+            {
+                @Override
+                public String getUserProperty( String key )
+                {
+                    return config.session.getUserProperties().get( key );
+                }
+
+                @Override
+                public Model getRawModel( Path p )
+                {
+                    return config.modelPool.get( p );
+                }
+
+                @Override
+                public Model getRawModel( String groupId, String artifactId )
+                {
+                    return config.modelPool.get( groupId, artifactId, null );
```

Comment:
Stupid question: Does this pool need to be synchronized?

Reply:
It has never been, see https://maven.apache.org/ref/3.6.3/maven-core/xref/org/apache/maven/project/ReactorModelPool.html
We should be safe because making the buildplan is done in a single thread. Once the buildplan is ready, the builder kicks in and can execute with multiple threads.